File: csharp/VoronoiHull.cs
using System;
using System.Collections.Generic;
using System.Numerics;

namespace VoronoiHullStandalone
{
    public readonly struct Tetrahedron
    {
        public readonly int A, B, C, D;
        public Tetrahedron(int a, int b, int c, int d) { A = a; B = b; C = c; D = d; }
    }

    /// <summary>Reusable, grow-only mesh storage. Only the first VertexCount and IndexCount entries are valid.</summary>
    public sealed class MeshBuffer
    {
        public Vector3[] Positions = Array.Empty<Vector3>();
        public Vector3[] Normals = Array.Empty<Vector3>();
        public int[] Indices = Array.Empty<int>();
        public int VertexCount { get; internal set; }
        public int IndexCount { get; internal set; }

        internal void Reserve(int vertices, int indices)
        {
            if (Positions.Length < vertices)
            {
                int capacity = Grow(Positions.Length, vertices);
                Array.Resize(ref Positions, capacity);
                Array.Resize(ref Normals, capacity);
            }
            if (Indices.Length < indices) Array.Resize(ref Indices, Grow(Indices.Length, indices));
        }

        private static int Grow(int oldCapacity, int needed)
        {
            int capacity = Math.Max(16, oldCapacity);
            while (capacity < needed) capacity = checked(capacity * 2);
            return capacity;
        }
    }

    /// <summary>Reusable topology and polygon workspace. Use one instance per concurrent build.</summary>
    public sealed class Scratch
    {
        internal readonly Dictionary<ulong, int> EdgeToFace = new Dictionary<ulong, int>();
        internal int[] EdgeA = Array.Empty<int>(), EdgeB = Array.Empty<int>();
        internal int[] Head = Array.Empty<int>(), Next = Array.Empty<int>(), TetIndex = Array.Empty<int>();
        internal Vector3[] Points = Array.Empty<Vector3>();
        internal float[] Angles = Array.Empty<float>();
        internal int FaceCount, NodeCount;

        internal void Reset(int maxEdges)
        {
            EdgeToFace.Clear();
            FaceCount = NodeCount = 0;
            if (EdgeA.Length < maxEdges)
            {
                Array.Resize(ref EdgeA, maxEdges);
                Array.Resize(ref EdgeB, maxEdges);
                Array.Resize(ref Head, maxEdges);
                Array.Resize(ref Next, maxEdges);
                Array.Resize(ref TetIndex, maxEdges);
            }
        }

        internal void ReservePoints(int count)
        {
            if (Points.Length >= count) return;
            int capacity = Math.Max(8, Points.Length);
            while (capacity < count) capacity = checked(capacity * 2);
            Array.Resize(ref Points, capacity);
            Array.Resize(ref Angles, capacity);
        }
    }

    public static class VoronoiHull
    {
        /// <summary>
        /// Builds the boundary between filled and empty sites from an existing 3D Delaunay
        /// tetrahedralization and its circumcenters. A face is emitted only when its filled
        /// endpoint is selected by emitCell (or for every filled endpoint when emitCell is null).
        /// Reuse scratch and mesh to avoid allocations after capacities have warmed up.
        /// </summary>
        public static MeshBuffer Build(
            Vector3[] sites, Tetrahedron[] tetrahedra, Vector3[] circumcenters,
            bool[] filled, Scratch scratch, MeshBuffer mesh, bool[] emitCell = null,
            float duplicateEpsilon = 1e-5f)
        {
            if (sites == null || tetrahedra == null || circumcenters == null || filled == null || scratch == null || mesh == null)
                throw new ArgumentNullException("Build inputs cannot be null");
            if (sites.Length != filled.Length || (emitCell != null && emitCell.Length != sites.Length) || tetrahedra.Length != circumcenters.Length)
                throw new ArgumentException("Input array lengths do not match");
            if (duplicateEpsilon < 0) throw new ArgumentOutOfRangeException(nameof(duplicateEpsilon));

            mesh.VertexCount = mesh.IndexCount = 0;
            scratch.Reset(checked(tetrahedra.Length * 6));
            for (int t = 0; t < tetrahedra.Length; t++)
            {
                Tetrahedron tet = tetrahedra[t];
                CheckSite(tet.A, sites.Length); CheckSite(tet.B, sites.Length);
                CheckSite(tet.C, sites.Length); CheckSite(tet.D, sites.Length);
                AddEdge(scratch, tet.A, tet.B, t); AddEdge(scratch, tet.A, tet.C, t);
                AddEdge(scratch, tet.A, tet.D, t); AddEdge(scratch, tet.B, tet.C, t);
                AddEdge(scratch, tet.B, tet.D, t); AddEdge(scratch, tet.C, tet.D, t);
            }

            float epsilonSq = duplicateEpsilon * duplicateEpsilon;
            for (int face = 0; face < scratch.FaceCount; face++)
            {
                int a = scratch.EdgeA[face], b = scratch.EdgeB[face];
                if (filled[a] == filled[b]) continue;
                int source = filled[a] ? a : b;
                if (emitCell != null && !emitCell[source]) continue;
                if (!IsClosed(scratch, face, tetrahedra)) continue; // unbounded Voronoi face

                Vector3 direction = sites[filled[a] ? b : a] - sites[source];
                float lengthSq = direction.LengthSquared();
                if (lengthSq <= 1e-20f) continue;
                Vector3 normal = direction / MathF.Sqrt(lengthSq);

                int count = 0;
                Vector3 centroid = Vector3.Zero;
                for (int node = scratch.Head[face]; node >= 0; node = scratch.Next[node])
                {
                    Vector3 point = circumcenters[scratch.TetIndex[node]];
                    if (!Finite(point)) continue;
                    bool duplicate = false;
                    for (int i = 0; i < count; i++)
                        if (Vector3.DistanceSquared(scratch.Points[i], point) <= epsilonSq) { duplicate = true; break; }
                    if (duplicate) continue;
                    scratch.ReservePoints(count + 1);
                    scratch.Points[count++] = point;
                    centroid += point;
                }
                if (count < 3) continue;
                centroid /= count;

                Vector3 tangent = Vector3.Normalize(Vector3.Cross(normal,
                    MathF.Abs(normal.X) < 0.9f ? Vector3.UnitX : Vector3.UnitY));
                Vector3 bitangent = Vector3.Cross(normal, tangent);
                for (int i = 0; i < count; i++)
                {
                    Vector3 delta = scratch.Points[i] - centroid;
                    scratch.Angles[i] = MathF.Atan2(Vector3.Dot(delta, bitangent), Vector3.Dot(delta, tangent));
                }
                // Insertion sort is efficient for the small polygons around Delaunay edges.
                for (int i = 1; i < count; i++)
                {
                    Vector3 point = scratch.Points[i]; float angle = scratch.Angles[i];
                    int j = i - 1;
                    while (j >= 0 && scratch.Angles[j] > angle)
                    {
                        scratch.Points[j + 1] = scratch.Points[j];
                        scratch.Angles[j + 1] = scratch.Angles[j];
                        j--;
                    }
                    scratch.Points[j + 1] = point; scratch.Angles[j + 1] = angle;
                }

                // A valid polygon can still contain nearly collinear circumcenters.
                float signedArea = 0;
                for (int i = 1; i < count - 1; i++)
                    signedArea += Vector3.Dot(Vector3.Cross(scratch.Points[i] - scratch.Points[0],
                        scratch.Points[i + 1] - scratch.Points[0]), normal);
                if (MathF.Abs(signedArea) <= epsilonSq) continue;
                if (signedArea < 0) Array.Reverse(scratch.Points, 0, count);

                int first = mesh.VertexCount;
                mesh.Reserve(checked(first + count), checked(mesh.IndexCount + (count - 2) * 3));
                for (int i = 0; i < count; i++)
                {
                    mesh.Positions[mesh.VertexCount] = scratch.Points[i];
                    mesh.Normals[mesh.VertexCount++] = normal;
                }
                for (int i = 1; i < count - 1; i++)
                {
                    mesh.Indices[mesh.IndexCount++] = first;
                    mesh.Indices[mesh.IndexCount++] = first + i;
                    mesh.Indices[mesh.IndexCount++] = first + i + 1;
                }
            }
            return mesh;
        }

        private static void CheckSite(int index, int count)
        {
            if ((uint)index >= (uint)count) throw new ArgumentOutOfRangeException(nameof(index), "Tetrahedron site index is out of range");
        }

        private static bool Finite(Vector3 p) =>
            !float.IsNaN(p.X) && !float.IsNaN(p.Y) && !float.IsNaN(p.Z) &&
            !float.IsInfinity(p.X) && !float.IsInfinity(p.Y) && !float.IsInfinity(p.Z);

        private static void AddEdge(Scratch s, int a, int b, int tet)
        {
            if (a == b) return;
            if (a > b) { int swap = a; a = b; b = swap; }
            ulong key = ((ulong)(uint)a << 32) | (uint)b;
            if (!s.EdgeToFace.TryGetValue(key, out int face))
            {
                face = s.FaceCount++;
                s.EdgeToFace.Add(key, face);
                s.EdgeA[face] = a; s.EdgeB[face] = b; s.Head[face] = -1;
            }
            int node = s.NodeCount++;
            s.TetIndex[node] = tet;
            s.Next[node] = s.Head[face];
            s.Head[face] = node;
        }

        // Every triangle containing this edge must belong to two tetrahedra.
        // Otherwise the dual Voronoi face extends to infinity and needs clipping.
        private static bool IsClosed(Scratch s, int face, Tetrahedron[] tetrahedra)
        {
            int a = s.EdgeA[face], b = s.EdgeB[face];
            for (int node = s.Head[face]; node >= 0; node = s.Next[node])
            {
                Tetrahedron t = tetrahedra[s.TetIndex[node]];
                if (!OtherVertices(t, a, b, out int c, out int d)) return false;
                if (CountThird(s, face, tetrahedra, c) != 2 || CountThird(s, face, tetrahedra, d) != 2)
                    return false;
            }
            return true;
        }

        private static int CountThird(Scratch s, int face, Tetrahedron[] tetrahedra, int third)
        {
            int count = 0;
            for (int node = s.Head[face]; node >= 0; node = s.Next[node])
            {
                Tetrahedron t = tetrahedra[s.TetIndex[node]];
                if (t.A == third || t.B == third || t.C == third || t.D == third) count++;
            }
            return count;
        }

        private static bool OtherVertices(Tetrahedron t, int a, int b, out int c, out int d)
        {
            c = d = -1; int count = 0;
            if (t.A != a && t.A != b) { c = t.A; count++; }
            if (t.B != a && t.B != b) { if (count++ == 0) c = t.B; else d = t.B; }
            if (t.C != a && t.C != b) { if (count++ == 0) c = t.C; else d = t.C; }
            if (t.D != a && t.D != b) { if (count++ == 0) c = t.D; else d = t.D; }
            return count == 2 && c != d;
        }
    }
}

File: typescript/voronoiHull.ts
/** Flat xyz positions, flat abcd tetrahedra, and one xyz circumcenter per tetrahedron. */
export interface VoronoiInput {
  sites: ArrayLike<number>;
  tetrahedra: ArrayLike<number>;
  circumcenters: ArrayLike<number>;
  filled: ArrayLike<number>; // 0 = empty, nonzero = filled
  emitCell?: ArrayLike<number>; // optional ownership mask for chunked output
  duplicateEpsilon?: number;
}

/** Reused output. Only positions/normals[0..vertexCount*3) and indices[0..indexCount) are valid. */
export class MeshBuffer {
  positions = new Float32Array(0);
  normals = new Float32Array(0);
  indices = new Uint32Array(0);
  vertexCount = 0;
  indexCount = 0;

  reserve(vertices: number, indices: number): void {
    const floats = vertices * 3;
    if (this.positions.length < floats) {
      const size = grow(this.positions.length, floats);
      const nextPositions = new Float32Array(size);
      const nextNormals = new Float32Array(size);
      nextPositions.set(this.positions);
      nextNormals.set(this.normals);
      this.positions = nextPositions;
      this.normals = nextNormals;
    }
    if (this.indices.length < indices) {
      const next = new Uint32Array(grow(this.indices.length, indices));
      next.set(this.indices);
      this.indices = next;
    }
  }
}

/** Reusable topology and polygon workspace. Use a separate instance for concurrent builds. */
export class Scratch {
  edgeA = new Int32Array(0);
  edgeB = new Int32Array(0);
  head = new Int32Array(0);
  next = new Int32Array(0);
  tetIndex = new Int32Array(0);
  faceCount = 0;
  nodeCount = 0;

  hashA = new Int32Array(0);
  hashB = new Int32Array(0);
  hashFace = new Int32Array(0);
  hashStamp = new Uint32Array(0);
  generation = 0;

  points = new Float64Array(0);
  angles = new Float64Array(0);

  reset(maxEdges: number): void {
    this.faceCount = this.nodeCount = 0;
    if (this.edgeA.length < maxEdges) {
      this.edgeA = new Int32Array(maxEdges);
      this.edgeB = new Int32Array(maxEdges);
      this.head = new Int32Array(maxEdges);
      this.next = new Int32Array(maxEdges);
      this.tetIndex = new Int32Array(maxEdges);
    }
    // Keep load under 50%. Generation stamps avoid clearing the hash table per run.
    const hashSize = nextPowerOfTwo(Math.max(16, maxEdges * 2));
    if (this.hashA.length < hashSize) {
      this.hashA = new Int32Array(hashSize);
      this.hashB = new Int32Array(hashSize);
      this.hashFace = new Int32Array(hashSize);
      this.hashStamp = new Uint32Array(hashSize);
      this.generation = 0;
    }
    this.generation = (this.generation + 1) >>> 0;
    if (this.generation === 0) {
      this.hashStamp.fill(0);
      this.generation = 1;
    }
  }

  reservePoints(count: number): void {
    if (this.angles.length >= count) return;
    const size = grow(this.angles.length, count);
    const nextAngles = new Float64Array(size);
    const nextPoints = new Float64Array(size * 3);
    nextAngles.set(this.angles);
    nextPoints.set(this.points);
    this.angles = nextAngles;
    this.points = nextPoints;
  }
}

function grow(oldSize: number, needed: number): number {
  let size = Math.max(16, oldSize);
  while (size < needed) size *= 2;
  return size;
}

function nextPowerOfTwo(value: number): number {
  let size = 1;
  while (size < value) size *= 2;
  return size;
}

function addEdge(s: Scratch, a: number, b: number, tet: number): void {
  if (a === b) return;
  if (a > b) { const swap = a; a = b; b = swap; }
  const mask = s.hashA.length - 1;
  let slot = (Math.imul(a, 0x9e3779b1) ^ Math.imul(b, 0x85ebca6b)) & mask;
  let face: number;
  while (true) {
    if (s.hashStamp[slot] !== s.generation) {
      face = s.faceCount++;
      s.hashStamp[slot] = s.generation;
      s.hashA[slot] = a;
      s.hashB[slot] = b;
      s.hashFace[slot] = face;
      s.edgeA[face] = a;
      s.edgeB[face] = b;
      s.head[face] = -1;
      break;
    }
    if (s.hashA[slot] === a && s.hashB[slot] === b) {
      face = s.hashFace[slot];
      break;
    }
    slot = (slot + 1) & mask;
  }
  const node = s.nodeCount++;
  s.tetIndex[node] = tet;
  s.next[node] = s.head[face];
  s.head[face] = node;
}

function countThird(s: Scratch, face: number, tets: ArrayLike<number>, third: number): number {
  let count = 0;
  for (let node = s.head[face]; node >= 0; node = s.next[node]) {
    const base = s.tetIndex[node] * 4;
    if (tets[base] === third || tets[base + 1] === third ||
        tets[base + 2] === third || tets[base + 3] === third) count++;
  }
  return count;
}

// A boundary triangle containing this edge means its dual face extends to infinity.
function isClosed(s: Scratch, face: number, tets: ArrayLike<number>): boolean {
  const a = s.edgeA[face], b = s.edgeB[face];
  for (let node = s.head[face]; node >= 0; node = s.next[node]) {
    const base = s.tetIndex[node] * 4;
    let c = -1, d = -1, count = 0;
    for (let i = 0; i < 4; i++) {
      const vertex = tets[base + i];
      if (vertex !== a && vertex !== b) {
        if (count++ === 0) c = vertex;
        else d = vertex;
      }
    }
    if (count !== 2 || c === d || countThird(s, face, tets, c) !== 2 ||
        countThird(s, face, tets, d) !== 2) return false;
  }
  return true;
}

/**
 * Triangulate bounded Voronoi faces between filled and empty sites. The input
 * tetrahedralization and circumcenters are supplied by the caller; no Delaunay
 * solver is included. Reuse scratch and mesh for allocation-free steady-state builds.
 */
export function buildVoronoiHull(input: VoronoiInput, scratch: Scratch, mesh: MeshBuffer): MeshBuffer {
  const { sites, tetrahedra: tets, circumcenters: centers, filled, emitCell } = input;
  if (sites.length % 3 || tets.length % 4 || centers.length !== tets.length / 4 * 3 ||
      filled.length !== sites.length / 3 || (emitCell && emitCell.length !== filled.length))
    throw new RangeError("Input array lengths do not match");
  const epsilon = input.duplicateEpsilon ?? 1e-5;
  if (epsilon < 0) throw new RangeError("duplicateEpsilon must be nonnegative");
  const epsilonSq = epsilon * epsilon;
  const siteCount = sites.length / 3;
  const tetCount = tets.length / 4;
  mesh.vertexCount = mesh.indexCount = 0;
  scratch.reset(tetCount * 6);

  for (let t = 0; t < tetCount; t++) {
    const i = t * 4;
    const a = tets[i], b = tets[i + 1], c = tets[i + 2], d = tets[i + 3];
    if (!Number.isInteger(a) || !Number.isInteger(b) || !Number.isInteger(c) || !Number.isInteger(d) ||
        Math.min(a, b, c, d) < 0 || Math.max(a, b, c, d) >= siteCount)
      throw new RangeError("Tetrahedron site index is out of range");
    addEdge(scratch, a, b, t); addEdge(scratch, a, c, t); addEdge(scratch, a, d, t);
    addEdge(scratch, b, c, t); addEdge(scratch, b, d, t); addEdge(scratch, c, d, t);
  }

  for (let face = 0; face < scratch.faceCount; face++) {
    const a = scratch.edgeA[face], b = scratch.edgeB[face];
    if (!!filled[a] === !!filled[b]) continue;
    const source = filled[a] ? a : b;
    if (emitCell && !emitCell[source]) continue;
    if (!isClosed(scratch, face, tets)) continue;

    const src = source * 3, dst = (source === a ? b : a) * 3;
    let nx = sites[dst] - sites[src], ny = sites[dst + 1] - sites[src + 1],
        nz = sites[dst + 2] - sites[src + 2];
    const normalLength = Math.hypot(nx, ny, nz);
    if (!(normalLength > 1e-10)) continue;
    nx /= normalLength; ny /= normalLength; nz /= normalLength;

    let count = 0, cx = 0, cy = 0, cz = 0;
    for (let node = scratch.head[face]; node >= 0; node = scratch.next[node]) {
      const ci = scratch.tetIndex[node] * 3;
      const x = centers[ci], y = centers[ci + 1], z = centers[ci + 2];
      if (!Number.isFinite(x) || !Number.isFinite(y) || !Number.isFinite(z)) continue;
      let duplicate = false;
      for (let i = 0; i < count; i++) {
        const p = i * 3, dx = scratch.points[p] - x, dy = scratch.points[p + 1] - y,
              dz = scratch.points[p + 2] - z;
        if (dx * dx + dy * dy + dz * dz <= epsilonSq) { duplicate = true; break; }
      }
      if (duplicate) continue;
      scratch.reservePoints(count + 1);
      const p = count++ * 3;
      scratch.points[p] = x; scratch.points[p + 1] = y; scratch.points[p + 2] = z;
      cx += x; cy += y; cz += z;
    }
    if (count < 3) continue;
    cx /= count; cy /= count; cz /= count;

    // tangent = normalize(cross(normal, X or Y)); bitangent = cross(normal, tangent)
    const ax = Math.abs(nx) < 0.9 ? 1 : 0, ay = ax ? 0 : 1;
    let tx = -nz * ay, ty = nz * ax, tz = nx * ay - ny * ax;
    const tangentLength = Math.hypot(tx, ty, tz);
    tx /= tangentLength; ty /= tangentLength; tz /= tangentLength;
    const bx = ny * tz - nz * ty, by = nz * tx - nx * tz, bz = nx * ty - ny * tx;
    for (let i = 0; i < count; i++) {
      const p = i * 3, x = scratch.points[p] - cx, y = scratch.points[p + 1] - cy,
            z = scratch.points[p + 2] - cz;
      scratch.angles[i] = Math.atan2(x * bx + y * by + z * bz, x * tx + y * ty + z * tz);
    }
    for (let i = 1; i < count; i++) {
      const p = i * 3, x = scratch.points[p], y = scratch.points[p + 1],
            z = scratch.points[p + 2], angle = scratch.angles[i];
      let j = i - 1;
      while (j >= 0 && scratch.angles[j] > angle) {
        const dstp = (j + 1) * 3, srcp = j * 3;
        scratch.points[dstp] = scratch.points[srcp];
        scratch.points[dstp + 1] = scratch.points[srcp + 1];
        scratch.points[dstp + 2] = scratch.points[srcp + 2];
        scratch.angles[j + 1] = scratch.angles[j];
        j--;
      }
      const dstp = (j + 1) * 3;
      scratch.points[dstp] = x; scratch.points[dstp + 1] = y; scratch.points[dstp + 2] = z;
      scratch.angles[j + 1] = angle;
    }

    let signedArea = 0;
    const x0 = scratch.points[0], y0 = scratch.points[1], z0 = scratch.points[2];
    for (let i = 1; i < count - 1; i++) {
      const p = i * 3, q = (i + 1) * 3;
      const ux = scratch.points[p] - x0, uy = scratch.points[p + 1] - y0, uz = scratch.points[p + 2] - z0;
      const vx = scratch.points[q] - x0, vy = scratch.points[q + 1] - y0, vz = scratch.points[q + 2] - z0;
      signedArea += (uy * vz - uz * vy) * nx + (uz * vx - ux * vz) * ny + (ux * vy - uy * vx) * nz;
    }
    if (Math.abs(signedArea) <= epsilonSq) continue;
    const first = mesh.vertexCount;
    mesh.reserve(first + count, mesh.indexCount + (count - 2) * 3);
    for (let i = 0; i < count; i++) {
      const from = (signedArea < 0 ? count - 1 - i : i) * 3;
      const to = mesh.vertexCount++ * 3;
      mesh.positions[to] = scratch.points[from];
      mesh.positions[to + 1] = scratch.points[from + 1];
      mesh.positions[to + 2] = scratch.points[from + 2];
      mesh.normals[to] = nx; mesh.normals[to + 1] = ny; mesh.normals[to + 2] = nz;
    }
    for (let i = 1; i < count - 1; i++) {
      mesh.indices[mesh.indexCount++] = first;
      mesh.indices[mesh.indexCount++] = first + i;
      mesh.indices[mesh.indexCount++] = first + i + 1;
    }
  }
  return mesh;
}
